DOCSP-15135 crud read sort update (#66)
mongodb/docs-java: source/fundamentals/builders/sort.txt

:java-docs:`Bson <apidocs/bson/org/bson/conversions/Bson.html>` objects. 
To learn more, see our :doc:`guide on builders </fundamentals/builders/>`.

You should read this guide if you would like to use builders to specify sort
criteria for your queries.

If you want to learn the fundamentals of sorting in the MongoDB Java
driver, consider reading our
:doc:`guide on sorting </fundamentals/crud/read-operations/sort/>`.

.. _sorts-builders-sort-example:

> The examples on this page use a sample collection, ``sort_example``, that

Reviewer: Do we ever use the collection name "sort_example" in any of the snippets? I wasn't able to find any instances on this page, so maybe we can omit the name and just refer to it as the example collection?

Author: Thanks for the note. You are right, it is not referred to. Fix: https://github.com/mongodb/docs-java/pull/66/commits/3273cb0926c61881026ca6d759d4520857076100

Going forward, I will try to be more aware of what needs an explicit name, and try to be more aware of when names are helpful/unhelpful.

> You should read this guide if you would like to:

Author: Thanks for the note. In a meeting, me and Chris decided that overview sections will have the formal structure of defining key terms, laying out what the article will say using the phrasingn "you should read this guide if you...", and then providing an alternate location to go to if this is not the right place. If you disagree with this structure, we should have a team meeting to decide on we will structure overview paragraphs going forward.

* Use builders to specify sort criteria for your queries.

> If you want to learn the fundamentals of sorting in the MongoDB Java

Author: Thanks for the note. This language was chosen in a 1-1 with Chris. The reason for it was that Overview sections should use an imperative tone sparingly. This section should help a reader decide if this page is for them, "This page will show you Z. if you want X, read this article".

> If you want to learn the fundamentals of sorting in the MongoDB Java

Reviewer: To be clear, I think both the suggestion and your current wording are technically imperative (i.e. giving the reader a command) language, so you're not reducing it either way. As you said, we want to use the overview section to help a reader decide where to go by presenting them with "if you want X, read [Y]." While most definitions of "imperative" direct you to its simplest form "Do X", if you dig further, you might stumble upon Kant's "hypothetical imperative":

> Hypothetical imperatives tell us how to act in order to achieve a specific goal and the commandment of reason applies only conditionally, e.g. "I must study to get a degree."

"See X to learn Y" and "If you want X, read [Y]" are equivalent, hypothetical imperatives.

driver, consider reading our
:doc:`guide on sorting </fundamentals/crud/read-operations/sort/>`.

.. warning::

   MongoDB does not provide guaranteed sort order in the event of **ties**. A
   tie occurs when two or more documents contain identical values in the
   field being used for a sort. For a detailed discussion of ties see the
   :ref:`handling ties section of our guide on sorting <sorts-crud-handling-ties>`

The Sorts Class
---------------

> The ``Sorts`` class is a builder that provides static factory methods for all sort

Author: See https://github.com/mongodb/docs-java/pull/66#discussion_r601822352

> criteria operators supported by MongoDB. These methods return a ``Bson`` object

Author: Thanks for the note. See https://github.com/mongodb/docs-java/pull/66#discussion_r601822352. Also, `Bson` object should link to the `Bson` interface API documentation.

that you can pass to the  ``sort()`` method of a ``FindIterable`` instance or to 
``Aggregates.sort()``. If you want to learn more about ``Aggregates``, see our 
:doc:`guide on the Aggregates builder </fundamentals/builders/aggregates>`.

.. _sorts-builders-sort-example:

> The following examples show you how to use the methods

Author: Thank you for the note. After rereading, I agree this sentence is unhelpful. I have removed the above sentence, and moved the sample collection out of the `Sorts Class` section.

Fix: https://github.com/mongodb/docs-java/pull/66/commits/098201d7ed944bacdcdbd7ec5853384d0c55693e

You should read this guide if you would like to:

* Use builders to specify sort criteria for your queries.

If you want to learn the fundamentals of sorting in the MongoDB Java
driver, consider reading our
:doc:`guide on sorting </fundamentals/crud/read-operations/sort/>`.

.. warning::

>    MongoDB does not provide guaranteed sort order in the event of **ties**. A

Reviewer 2: You can often replace verb + past participle adjective with the verb:
```suggestion
   MongoDB does not guarantee sort order in the event of **ties**. A
```

Author: Thank you for the note.

Going forward, I will watch out for verb + past participle usage. Can this check be automated?

   tie occurs when two or more documents contain identical values in the

>    field being used for a sort. For a detailed discussion of ties see the

Reviewer 2: "Field being used for a sort" feels a bit clunky. Maybe "in the sorted field"?

Author: Thanks for the note. I agree my current language is clunky. "In the sorted field" sounds slightly misleading to me as I fear a user could mistake that to mean the "fields are being sorted" or "only some fields get sorted".  What about "field being used to order your results"?

Change: https://github.com/mongodb/docs-java/pull/66/commits/ce1fe60d3bd9e8acad5e6b37898d73f800747796

Reviewer 2: Sure, sounds good.

>    field being used for a sort. For a detailed discussion of ties see the

Reviewer 2: ```suggestion
   field being used for a sort. For a detailed discussion of ties, see the
```

Author: Thank you for pointing this out. Need to come up with a better system for grammar checking.

>    :ref:`handling ties section of our guide on sorting <sorts-crud-handling-ties>`

Reviewer 2: ```suggestion
   :ref:`handling ties section of our guide on sorting <sorts-crud-handling-ties>`.
```

Author: Thank you for pointing this out. Going forward, I need to figure out a system for catching these types of punctuation errors.

You should read this guide if you would like to:

> * Use builders to specify sort criteria for your queries.

Author: Thanks for the note. Fix: https://github.com/mongodb/docs-java/pull/66/commits/164972f20dc5d3c875fca859681da60ebad0a8de

Going forward, I will avoid single item lists. I will try to refine my ability to correctly use formatting conventions.

* Perform ascending sorts and descending sorts.
* Combine sort criteria.
* Sort on the text score of a
  :manual:`text search </core/text-search-operators/>`.

If you want to learn the fundamentals of sorting in the MongoDB Java
driver, consider reading our
:doc:`guide on sorting </fundamentals/crud/read-operations/sort/>`.

The Sorts Class
---------------

The :java-docs:`Sorts </apidocs/mongodb-driver-core/com/mongodb/client/model/Sorts.html>`
class is a builder that provides static factory methods for all sort criteria
operators supported by MongoDB. These methods return a
:java-docs:`Bson </apidocs/bson/org/bson/conversions/Bson.html>`
object that you can pass to the 
:java-docs:`sort() </apidocs/mongodb-driver-sync/com/mongodb/client/FindIterable.html#sort(org.bson.conversions.Bson)>`
method of a ``FindIterable`` instance or to 
:java-docs:`Aggregates.sort() </apidocs/mongodb-driver-core/com/mongodb/client/model/Aggregates.html#sort(org.bson.conversions.Bson)>`.

> If you want to learn more about ``Aggregates``, see our 

Author: Thank you for the note. I agree, `Aggregates` requires more context. Fix:

Going forward, I will try to refine this area of reader empathy, and not count on internal formatting conventions to effectively convey meaning.

>    field being used to order your results. For a detailed discussion of ties, 

Author: Thanks for the note. I added the warning because tied sort behavior seemed to me the main thing about sorting that could trip up a developer in their app, and they have skimmed the intro and not be aware of the other article. 

This probably is assuming too much however, and focusing this guide around the builder and less around sorting is probably for the best. I will remove the warning. Fix:

   {"_id": 2, "letter": "a", "food": "donuts and coffee"}
   {"_id": 6, "letter": "c", "food": "maple donut"}

The Sorts Class
---------------

The ``Sorts`` class is a builder that provides static factory methods for all sort
criteria operators supported by MongoDB. These methods return a ``Bson`` object
that you can pass to the  ``sort()`` method of a ``FindIterable`` instance or to 
``Aggregates.sort()``. If you want to learn more about the ``Aggregates``
class, see our 
:doc:`guide on the Aggregates builder </fundamentals/builders/aggregates>`.

For more information about the classes and interfaces in this section, see the
following API documentation:
   
- :java-docs:`Sorts </apidocs/mongodb-driver-core/com/mongodb/client/model/Sorts.html>` 
- :java-docs:`Bson </apidocs/bson/org/bson/conversions/Bson.html>`
- :java-docs:`FindIterable </apidocs/mongodb-driver-sync/com/mongodb/client/FindIterable.html>`

> - :java-docs:`Aggregates </apidocs/mongodb-driver-core/com/mongodb/client/model/Aggregates.html>`.

Reviewer: ```suggestion
- :java-docs:`Aggregates </apidocs/mongodb-driver-core/com/mongodb/client/model/Aggregates.html>`
```

   {"_id": 2, "letter": "a", "food": "donuts and coffee"}
   {"_id": 6, "letter": "c", "food": "maple donut"}

The Sorts Class
---------------

The ``Sorts`` class is a builder that provides static factory methods for all sort
criteria operators supported by MongoDB. These methods return a ``Bson`` object
that you can pass to the  ``sort()`` method of a ``FindIterable`` instance or to 
``Aggregates.sort()``. If you want to learn more about ``Aggregates``, see our 
:doc:`guide on the Aggregates builder </fundamentals/builders/aggregates>`.

> The following two code snippets are equivalent, however we recommend you use the

Reviewer 2: Avoid joining two sentences with a comma:
```suggestion
The following two code snippets are equivalent. However, we recommend you use the
```

Author: Thank you for the suggestion. I am not sure I agree with this edit, as using a period makes me read it as two separate thoughts (where in my head it is one thought "code snippets do the same thing but one is preferred"). Will commit suggestion, but would appreciate follow up discussion.

``Sorts`` builder to specify your sort criteria:

.. code-block:: java

   import static com.mongodb.client.model.Sorts.ascending;

   // <MongoCollection setup code here>
   

>    //Recommended

Reviewer 2: Purely personal preference:
```suggestion
   // Recommended
```

Author: Thank you for the note. This is the style we have been using as well. I think I need to update my Intellij settings to automatically lead code comments with spaces.

   collection.find().sort(ascending("_id"));

.. code-block:: java

   // <MongoCollection setup code here>
   

>    //Not Recommended

Reviewer 2: ```suggestion
   // Not Recommended
```

   collection.find().sort(new Document("_id", 1));

> The code snippets above both sort the documents in the

Reviewer 2: The "both" here feels a bit redundant. You could also say "Both code snippets..."
```suggestion
The code snippets above sort the documents in the
```

Author: Thank you for the note.  I agree this phrasing could be improved. think I prefer "Both code snippets", as in my head it stresses equivalence more heavily.

```suggestion
Both code snippets above sort the documents in the
```

   {"_id": 2, "letter": "a", "food": "donuts and coffee"}
   {"_id": 6, "letter": "c", "food": "maple donut"}

The Sorts Class
---------------

The ``Sorts`` class is a builder that provides static factory methods for all sort
criteria operators supported by MongoDB. These methods return a ``Bson`` object
that you can pass to the  ``sort()`` method of a ``FindIterable`` instance or to 
``Aggregates.sort()``. If you want to learn more about ``Aggregates``, see our 
:doc:`guide on the Aggregates builder </fundamentals/builders/aggregates>`.

The following two code snippets are equivalent. However, we recommend you use the
``Sorts`` builder to specify your sort criteria:

.. code-block:: java

   import static com.mongodb.client.model.Sorts.ascending;

   // <MongoCollection setup code here>
   
   // Recommended
   collection.find().sort(ascending("_id"));

> .. code-block:: java

Author: Thanks for the note. I have removed the equivalency example -- see https://github.com/mongodb/docs-java/pull/66#discussion_r602642042

   // <MongoCollection setup code here>
   
   // Not Recommended
   collection.find().sort(new Document("_id", 1));

Both code snippets above sort the documents in the
:ref:`sort_example collection <sorts-crud-sort-example>`  

> in the following order: 

Author: Thanks for the note. I am removing the equivalency example as suggested. I think it is getting too out of scope to discuss `Bson` documents in any depth, and most likely isn't all that helpful to readers. Fix: https://github.com/mongodb/docs-java/pull/66/commits/a21b7a8b39f91a627142ef7d8bfda65b1ee8ace5

   {"_id": 2, "letter": "a", "food": "donuts and coffee"}
   {"_id": 6, "letter": "c", "food": "maple donut"}

> For more information about the classes and interfaces in this section, see the

Author: Thank you for the note. This phrasing was modeled after Chris Cho's phrasing [here ](https://docs-mongodbcom-staging.corp.mongodb.com/398c0f5/java/docsworker-xlarge/DOCSP-12161-fundamentals-codecs/fundamentals/data-formats/codecs/#codec). See [here](https://github.com/mongodb/docs-java/pull/66#discussion_r601822352) for response to other comment.

   
   collection.find().sort(descending("_id"));

> The output of the example above should look something like this: 

Reviewer 2: You can always tidy up a sentence by moving the subject around and using verbs to convey the other relevant information:
```suggestion
The above example should output something like this: 
```

Author: Thank you for the note. I agree that phrasing is much better. I think I got too into the habit of saying "above should look something like this". Will try to reuse text more carefully.

   {"_id": 4, "letter": "b", "food": "coffee with sugar"}
   ...

Combining Sort Criteria
-----------------------

To combine sort criteria, use the ``Sorts.orderBy()`` static factory
method. The ``orderBy()`` method builds sort criteria that apply passed 
in sort criteria from left to right in the event of ties.

The following example sorts the documents in the 
:ref:`sort_example collection <sorts-builders-sort-example>` in descending order

> on the ``letter`` field and ascending order on the ``_id`` field:

Author: Thank you for the note. I agree adding "and in the event of a tie" makes the sentence more clear. Fix:

Going forward, I will try to improve my ability to hear these types of opportunities to improve the clarity or sentences.

   {"_id": 3, "letter": "a", "food": "maple syrup"}
   {"_id": 5, "letter": "a", "food": "milk and cookies"}

Text Search
-----------

You can specify the order of the results of a 
:manual:`text search </text-search/>` by how closely they match your
search string. Each of your search results has a
:manual:`text score </reference/operator/aggregation/meta/#exp._S_meta>`, a
numerical value indicating how well that result matches your search. 
Use the ``Sorts.metaTextScore()`` static factory method to build your sort
criteria to sort by the text score.

.. warning:: Make Sure to Create a Text Index

   You need a :manual:`text index </core/index-text/>` on your collection to perform a text search. See the server manual documentation for more
   information on how to

>    :manual:`create a text index </core/index-text/#create-text-index>`.

Reviewer 3: Suggestion: I think "create a text index" should link to the builders > indexes page where it creates a text index

Author: Thank you for the note. The [Text Indexes Section](https://docs-mongodbcom-staging.corp.mongodb.com/86d2198/java/docsworker-xlarge/DOCSP-15135-CRUD-Read-Sort-Update/fundamentals/builders/indexes/#text-indexes) does not provide a description of what a text index is used for. My goal in linking to the server documentation is to help readers gain an understanding of how text indexes are used in case they are not familiar with them, and I do not believe the Text indexes section provides that information. Please let me know if you disagree with this assessment.

Reviewer 3: Agreed, but is there a reason you're linking to the same page twice?  I think the first sentence on its own is good enough in that case.

Author: One link describes what a text index is, the other describes how to create one. Although they are both on the same page, I think having both gives readers a way to understand what I am describing (what is a text index?) and how to proceed given the situation (how do I make one?).

   {"_id": 3, "letter": "a", "food": "maple syrup"}
   {"_id": 5, "letter": "a", "food": "milk and cookies"}

Text Score
----------

To specify a sort by the text score of a text search, use the

> ``Sorts.metaTextScore()`` static factory method. For a detailed usage example

Author: Thanks for the note. I changed this to just "for a detailed example". Going forward, I will try to avoid overloading terms in the documentation. Fix: https://github.com/mongodb/docs-java/pull/66/commits/4ca761f59685fc5ab3e5adc47d36cd702f3548bf
